lib/transport/datagouvfr/client/datasets.ex

```diff
@@ -3,7 +3,50 @@ defmodule Transport.Datagouvfr.Client.Datasets do
   A client to manipulate https://www.data.gouv.fr/api/1/datasets endpoints
   """
 
-  import Transport.Datagouvfr.Client, only: [get_request: 2, put_request: 3]
+  import TransportWeb.Gettext
+  import Transport.Datagouvfr.Client, only: [get_request: 2, put_request: 3,
+                                             post_request: 3, post_request: 4]
+
+  use Vex.Struct
+  alias __MODULE__
+
+  defstruct [description: nil,
+             frequency: nil,
+             licence: "ODbL",
+             organization: nil,
+             title: nil]
+
+  validates :description,
+               length: [min: 1,
+                        message: dgettext("dataset", "A description is needed")]
+
+  validates :frequency,
+               inclusion: ["bimonthly", "fourTimesAWeek", "semidaily", "weekly",
+                           "threeTimesAWeek", "hourly", "monthly", "unknown",
+                           "semimonthly", "threeTimesAMonth", "quarterly",
+                           "triennial", "threeTimesADay", "punctual",
+                           "continuous", "quinquennial", "semiweekly",
+                           "biweekly", "threeTimesAYear", "irregular", "annual",
+                           "semiannual", "daily", "biennial", "fourTimesADay"]
+
+  validates :organization,
+               length: [min: 1,
+                        message: dgettext("dataset", "An organization is needed")]
+
+  validates :title,
+               length: [min: 1,
+                        message: dgettext("dataset", "A title is needed")]
+
+  @spec new(map()) :: %__MODULE__{}
+  def new(%{} = map) do
+    map
+    |> Map.take(Enum.map(Map.keys(%Datasets{}), &Atom.to_string/1))
```

Comment:
```elixir
map
|> Map.take(Enum.map(keys(), &Atom.to_string/1))
#...

defp keys do
  __struct__
  |> Map.from_struct
  |> Map.keys
end
```

Otherwise, you're also mapping ```:__struct__```.